### ISSUE TITLE
Add option for only building listed tests and benchmarks

### ISSUE THREAD

**Issue by joshlf** [rust-lang/cargo#4014]
I have a crate with a very large number of tests and benchmarks, and as a result, doing `cargo test --no-run` takes ~2 minutes and doing `cargo bench --no-run` takes ~4 minutes on my MacBook.

It'd be great if there were a flag that caused only the tests/benchmarks which will be run to be built, which could greatly reduce build times if only a small subset of tests/benchmarks are being run.

For example, running `cargo test --omit-not-run foo` would disable any test/benchmark function not matching `foo`, and thus it wouldn't be included in the build. `cargo test --omit-not-run` (without test name arguments) would probably be equivalent to `cargo test` since all tests are run. The same behavior would hold for `cargo bench`.

**Comment by alexcrichton** [comment]
Right now Cargo supports filters like `--test foo` and `--bench bar`, would that work in this situation?

**Comment by joshlf** [comment]
I've never gotten those flags to work - no matter what I pass to them, I get `error: no bench target named ...`.

**Comment by alexcrichton** [comment]
Ah yeah they're intended for the Cargo targets of tests/benches, not the names in the tests themselves. Unfortunately the compiler provides no ability to compile just one function in a test file.

**Comment by joshlf** [comment]
Ah, gotcha. Is there any mechanism to instruct `rustc` to omit a particular AST item as if it were preceded by `#[cfg(false)]` or something like that (for posterity, I already asked about that [here](https://users.rust-lang.org/t/disable-block-using-rustc/10701))?

I also considered writing a compiler plugin, although I'm not sure if compiler plugins have enough information exposed to them to make that possible.

**Comment by alexcrichton** [comment]
Nah unfortunately short of spraying `#[cfg]` everywhere unconditionally there's no way to strip out ASTs at compile time.

Plugins, however, may have the power to tweak the ast! They're just unstable :(

**Comment by joshlf** [comment]
OK gotcha. Well, good to know! Do you think that a "disable this item" CLI flag (or maybe, "disable all items matching this pattern/rule/regex/etc) for `rustc` would be something that might be considered? I'd be happy to file an issue for it.

**Comment by joshlf** [comment]
Oh, and I think the big issue with a compiler plugin is being able to pass it the relevant information about /which/ tests to disable (e.g., is there a way for it to read `rustc` CLI flags?).

**Comment by alexcrichton** [comment]
I think it'd be awesome to prototype that as a plugin (if possible) and if it works could get some good traction for adding to rustc itself!

**Comment by joshlf** [comment]
Sounds good; I'll see what I can do!

**Comment by joshlf** [comment]
OK so I wrote a prototype compiler plugin to disable arbitrary pieces of code, but unfortunately [it looks like](https://users.rust-lang.org/t/modify-ast-after-macro-expansion-in-compiler-plugin/10736) my pass over the AST happens prior to macro expansion, and I'm not sure if there's a way around this limitation. Personally, I use macros a lot to define tests, so while the plugin works for tests that exist in the source code, it doesn't work for macro-generated tests.

However, while working on the plugin, I poked through `librustc_driver`, and it looks like modifying `rustc` so that the test harness only calls particular test functions should be pretty straightforward. I'd be happy to throw up a PR if there's interest, although I'm not sure if there's some more official process that I need to go through for that.
